### DISCUSSION TITLE
Zeroconf mode - how it work exactly?

### DISCUSSION THREAD

**Discussion by vampywiz17** [hsaturn/TinyMqtt#20] [General]
Hello,

i would like to ask some thing with zeroconf/mesh mode.

1. How it work "master" broker exactly? It get a virtual IP address and other systems like HA connect this address?
2. how much time that if the mesh network "degraded" to regenerate ourself? (chose other master, etc...)

Now that is all, that i would lime to ask, but possible that in will do others if no problem :)

**Comment by hsaturn** [comment]
Hello Vampy

Main idea behind zeroconf is to be hardware failure prone.
-> if the "declared" main tiny broker fails
-> if internet is down (if an external broker is used)
-> if the server of the house fails
-> etc..

In those case, I still want that Esp's topics are correctly sent/received.

Imagine a house with
1 - A main server with Mqtt Broker installed and Gui / internet server / Android server (to pilot your home with your phone)
2 - Many ESP with TinyMqtt everywhere
3 - One ESP (sensor) pusblish the temperature of a room
4 - Some ESP (controllers) controls the heating device(s) 

Sensor publish the topic "bedroom/temperature" while controllers are subsribed to that topic.

If the main server has a failure, nothing will work in the house. Heaters are off, lights are off etc.. and nothing can be controlled anymore.

With the mesh, things continue to work, in a degraded mode, but work ! You still can send the topic "kitchen/light/on" or "kitchen/light/off".

Even if some ESP are off (for example a fuse burnt), one will still claim to be the master.

Of course, you won't be able to control things from your phone but at least, you can switch lights. (I'm intending to replace all my switches by ESP+Tiny)

-----

The zeroconf is not yet working, nor it is developped, but the mesh has some limited features that are working.

Yet, I'm essentially working on bugs and unit testing to ensure the library won't become unstable => I really want to avoid at max the risk of regression  when releasing.

Once done, I'll work again on the zeroconf.

My ideas are
When a tiny broker is connected to another broker (tiny or not) it does not route topics but let the master do that for him.
When a tiny broker loose a connection to the master, then it tries to connect to a tiny broker of the subnetwork.
When this is not possible then the broker tries to promot itself as a broker
Naturally, other tiny brokers, because they are trying over the whole subnet will find it.

Yet, I know this will probably hard to implement, and many problems will occur.
The worst would be to have two promoted brokers, and thus two separated meshes. unable to communicate together.

Best regards
